### PR TITLE
Removal of snapshot with corrupt replica crashes osd

### PR DESCRIPTION
**Pull request by dzafman**
This is only a partial fix for http://tracker.ceph.com/issues/23875

### CODE REVIEW

**Comment by dzafman** [comment]
@liewegas @jdurgin @tchaikov @neha-ojha I created this pull request to start the discussion on taking some asserts out.

File: qa/standalone/scrub/osd-scrub-snaps.sh [hunk 1]
```diff
@@ -90,57 +90,58 @@ function create_scenario() {
     # Don't need to use ceph_objectstore_tool() function because osd stopped
 
     JSON="$(ceph-objectstore-tool --data-path $dir/${osd} --head --op list obj1)"
-    ceph-objectstore-tool --data-path $dir/${osd} "$JSON" --force remove
+    ceph-objectstore-tool --data-path $dir/${osd} "$JSON" --force remove || return 1
```
Comment (gregsfortytwo) [inline qa/standalone/scrub/osd-scrub-snaps.sh]: Why do we want to tolerate an error in every one of these? Do we expect stale data we might need to clean up?

Comment (dzafman) [reply qa/standalone/scrub/osd-scrub-snaps.sh]: Actually, this change causes the script to not tolerate errors.  Before I just assumed that these commands can not fail, so didn't bother to check.

**Comment by dzafman** [comment]
@gregsfortytwo Even with this change a replica can still crash when removing snapshots with certain corruptions.  It isn't clear that we shouldn't consider some corruptions as fatal anyway. 

```
    # When removing snapshots with a corrupt replica, it crashes.
    # See http://tracker.ceph.com/issues/23875
    if [ $which = "primary" ];
    then
        for i in `seq 1 7`
        do
            rados -p $poolname rmsnap snap$i
        done
    fi
```

File: qa/standalone/scrub/osd-scrub-snaps.sh (hunk 1 above)
Comment (gregsfortytwo) [reply qa/standalone/scrub/osd-scrub-snaps.sh]: Oh dur.
👍

**Comment by tchaikov** [comment]
will review this PR early tomorrow.

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -1981,25 +1981,36 @@ int do_meta(ObjectStore *store, string object, Formatter *formatter, bool debug,
   return 0;
 }
 
+enum rmtype {
+  STANDARD,
```
Comment (tchaikov) [inline src/tools/ceph_objectstore_tool.cc]: so

- STANDARD for removing both snap from snapmapper and the object
- SNAPMAP for removing the snap from snapmapper only
- NOSNAPMAP for remove the object only

i'd suggest rename `STANDARD` to `BOTH`, `NOSNAPMAP` to `OBJECT`, like
```c++
enum class Remove {
  BOTH,
  SNAPMAP,
  OBJECT
};

and update `remove_object()` like,
```c++
if (type == Remove::BOTH || type == Remove::SNAPMAP) {
  // remove snap from snapmap
}
if (type == Remove::BOTH || type == Remove::OBJECT) {
  // remove the object
}
```

easier to digest this way, IMHO.

**Comment by dzafman** [comment]
@tchaikov I rebased this pull request and made the change you requested.

**Comment by dzafman** [comment]
retest this please

**Comment by dzafman** [comment]
retest this please

**Comment by gregsfortytwo** [comment]
@dzafman I think you need to rebase this on master again to get in the WITH_SEASTAR fix.

**Comment by tchaikov** [comment]
@gregsfortytwo the jenkins always tries to rebase a tested PR against the latest master at that moment. so what we need is just:

retest this please

=)

**Comment by dzafman** [comment]
osd-scrub-snaps.sh isn't passing due to ceph-osdomap-tool crashing possibly because of e406d8eb9e1deb801ecb346169eaaf96adbb4b53

**Comment by jdurgin** [comment]
@dzafman is this ready to go after a rebase?

**Comment by gregsfortytwo** [comment]
@dzafman, I gather we should go through and re-review this now?

Note that the final FIXUP commit is missing a signoff statement.

**Comment by dzafman** [comment]
retest this please

**Comment by dzafman** [comment]
Rados test suite passed with unrelated failures:
dzafman-2018-11-08_20:13:57-rados-wip-zafman-testing-22476-distro-basic-smithi
	3238374 http://tracker.ceph.com/issues/20798. [ FAILED ] LibRadosLockECPP.LockExclusiveDurPP
	3238479 /usr/include/rados/buffer.h:657:61: error: expected string-literal before ‘)’ token
	3238552 scrub_test: assert len(objs['inconsistents']) == 1
	3238596 rados/test_librados_build.sh: /usr/bin/x86_64-linux-gnu-ld: cannot find -lradospp failed

**Comment by dzafman** [comment]
@jdurgin @gregsfortytwo @tchaikov This is ready for a final review and merge.  I rebased, made changes suggested by Kefu and passed the rados suite.

**Pull request merged**